Added support for a cover page when generating a PDF (#2004)

icnocop opened the pull request.

bitbonk commented:
Usually (for us) a cover page does not contain a lot of text but is more of a page with special layout, e.g.  A *big* title centered in the middle, a smaller subtitle, a company logo, no footers no headers, maybe a date and copyright notice in a small special font.

So the markdown is not really the problem. The bigger problem is: How can I define that special layout for the cover page alone? How can I have footers or headers with page numbering on all of my pages but not on the cover page (and the ToC pages)?

Does this PR make those things easier or even possible?

icnocop commented:
Hi @bitbonk.

I think footers, headers, page numbers, and having the ability to provide a different layout for different pages are different issues.

superyyrrzz commented:
I think one point from @bitbonk is that is is not flexible to use a Markdown file to define a cover page. IMO, a cover page is like some title in the middle of a page, but we can only get an upper-left H1 header by a Markdown input.

With this, I'd prefer a `_cover.pdf` instead of `cover.md`. DocFX can insert this PDF at the start of generated PDF, and users can put whatever in this one page `_cover.pdf`.

icnocop commented:
@bitbonk mentioned 

> the markdown is not really the problem

I understood that as "Markdown is fine." Was I mistaken?

In any case, markdown is flexible enough so we're not limited to only an upper-left H1 header.

For example, this markdown file will generate a page with "Cover Page Title" in the middle of the page:
```
<style>
.cover-title {
    height: 200px;
    width: 400px;
    position: fixed;
    text-align: center;
    top: 50%;
    left: 50%;
    margin-top: -100px;
    margin-left: -200px;
}
</style>
<div class="cover-title">Cover Page Title</div>
```

With these changes, if you name the file `cover.md` and include it in `docfx.json` as `content`, it will be added as the first page in the generated PDF.

Requiring a PDF file as the cover page doesn't seem very flexible or document-author-friendly to me.
It introduces another file format as input which is harder to diff and modify than the markdown and yaml files currently being used as inputs for conceptual documentation.

icnocop commented:
In regards to changing page margins, I think I agree with @PhilterPaper's comment here: https://github.com/wkhtmltopdf/wkhtmltopdf/issues/3635#issuecomment-360204672

> To remove or change margins on an arbitrary interior page would probably better be done with embedded directives (special HTML comments)

In that way, each page could have it's own margins defined from within the page itself.

We may want to also think about that approach for overriding the headers and footers in specific pages for example.

Instead of HTML comments, it may be better if they were "markdown comments".
For example, "markdown comments" can be written like the examples here: https://stackoverflow.com/a/20885980/90287.
`Microsoft.DocAsCode.HtmlToPdf` would probably have to parse the files to extract those directives though.

So basically, instead of littering `docfx.json` with options for each page, options could be defined within the pages themselves to override the global/default options.

icnocop commented:
Another approach other than using "markdown comments" is to have an accompanying JSON or YAML file for each markdown file that contains these properties.

For example cover.md.json:
```
{
  margin-bottom: 0,
  margin-left: 0,
  margin-right: 0,
  margin-top: 0,
  orientation: "Landscape",
  footer-center: "",
  footer-left: "",
  footer-right: "",
  header-center: "",
  header-left: "",
  header-right: ""
}
```
Those options would correspond to wkhtmltopdf's page, header, and footer options.
https://wkhtmltopdf.org/usage/wkhtmltopdf.txt

However, wkhtmltopdf doesn't currently support setting different margins and orientation for different pages.
See https://github.com/wkhtmltopdf/wkhtmltopdf/issues/3635 and https://github.com/wkhtmltopdf/wkhtmltopdf/issues/1564

superyyrrzz commented:
@icnocop Thank you for your thoughts. It convinced me that PDF is not an ideal cover input.

My other concern is whether we need `cover.md`. If it is built into HTML by docfx, put to output folder and added to `manifest.json`, that means `cover.html` is part of the output that users can view final site. However, `cover.html` should just be an intermediate output used by PDF pipeline. It should be excluced from output.

Inspired by your example, I think `cover.html` could be a purer input. wkhtml can use it as input directly, without need to borrow Markdown pipeline and pollute HTML output. It is more straightforward for cover page author to write styles in HTML than in Markdown.

bitbonk commented:
I think I would like it better if I just could provide a cover.html (that uses the styles of my PDF DocFX template) or even a cover.pdf instead of cover.md.

There wouldn’t be much text I‘d add to a cover page, hence the cover.md is not very useful. In my use cases, the cover page is more about layouting and less about (text)content.

superyyrrzz commented on Documentation/tutorial/docfx.exe_user_manual.md:
```diff
@@ -158,6 +158,8 @@ docfx pdf [<config_file_path>] [-o:<output_path>]
 
 Current design is that each TOC file generates a corresponding PDF file. Walk through [Walkthrough: Generate PDF Files](walkthrough/walkthrough_generate_pdf.md) to get start.
 
+If `cover.md` is found in a folder, it will be rendered as the cover page.
```
This document may also need update:
https://dotnet.github.io/docfx/tutorial/walkthrough/walkthrough_generate_pdf.html#step1-add-a-tocyml-specific-for-pdf

> TOC is also used as the cover page of the PDF

superyyrrzz commented on src/Microsoft.DocAsCode.HtmlToPdf/ConvertWrapper.cs:
```diff
@@ -132,6 +132,19 @@ private void ConvertCore(string basePath)
                             AbsolutePathInTocPageFileTransformer(tocPageFilePath);
                             htmlModels.Insert(0, new HtmlModel { Title = _pdfOptions.TocTitle, HtmlFilePath = tocPageFilePath });
                         }
+
```
DocFX generates one PDF for each TOC. So for a single TOC here, we need not find all the cover pages in manifest. We can make this conversion: only the `cover.xxx` that sibling to TOC in the input folder is treated as current PDF's cover page. Then the logic here is like `var coverPage = FindSiblingPage(tocManifestItem)`

superyyrrzz commented on src/Microsoft.DocAsCode.HtmlToPdf/ConvertWrapper.cs:
```diff
@@ -132,6 +132,19 @@ private void ConvertCore(string basePath)
+                        foreach (ManifestItem coverPage in coverPages)
+                        {
+                            var coverPageFilePath = Path.Combine(basePath, coverPage.SourceRelativePath);
+                            var coverPageHtmlFileName = $"{Path.GetFileNameWithoutExtension(coverPageFilePath)}.html";
```
DocFX can use [src and dest in file mapping](https://dotnet.github.io/docfx/tutorial/docfx.exe_user_manual.html#41-array-format) to make output path different with input path. So the accurate logic to get output HTML path is like:

    var coverPageHtmlFilePath = coverPage.OutputFiles["html"].RelativePath

BTW, we shall first determine if we need to add cover page into manifest.

superyyrrzz commented on src/Microsoft.DocAsCode.HtmlToPdf/PdfOptions.cs:
```diff
@@ -41,13 +41,18 @@ public class PdfOptions
         /// Gets or sets the "Table of Contents" bookmark title.
         /// </summary>
-        public string TocTitle { get; set; } = "Cover Page";
+        public string TocTitle { get; set; } = "Table of Contents";
```
I'd expect to keep the default value as "Cover Page", if user does not explicit provide a cover page. We can have a logic to change default value to "Table of Contents" if user provided cover page is found.

superyyrrzz commented on src/Microsoft.DocAsCode.HtmlToPdf/HtmlToPdfConverter.cs:
```diff
@@ -51,23 +51,30 @@ public HtmlToPdfConverter(IList<HtmlModel> htmlModels, HtmlToPdfOptions htmlToPd
         #region Public Methods
 
-        public IDictionary<string, int> GetHtmlToPdfNumberOfPages(IList<string> htmlFilePaths)
+        public IList<PartialPdfModel> GetHtmlToPdfNumberOfPages(IList<string> htmlFilePaths)
```
nit: update the method name, maybe `GetPartialPdfModels`

icnocop commented:
> My other concern is whether we need cover.md. If it is built into HTML by docfx, put to output folder and added to `manifest.json`, that means `cover.html` is part of the output that users can view final site.

I think it should be recommended that `cover.md` be placed in the `pdf` folder, like the `toc.yml` file specific for PDF.
https://dotnet.github.io/docfx/tutorial/walkthrough/walkthrough_generate_pdf.html#step1-add-a-tocyml-specific-for-pdf

In this way, the generated `cover.html` isn't part of the generated web site, and only part of the generated PDF.

> Inspired by your example, I think _cover.html could be a purer input. wkhtml can use it as input directly, without need to borrow Markdown pipeline and pollute HTML output. It is more straightforward for cover page author to write styles in HTML than in Markdown.

> I think I would like it better if I just could provide a cover.html (that uses the styles of my PDF DocFX template)

If you start from scratch, using the default PDF template, and want to create an HTML file for the cover page, you don't have access to the CSS styles as part of the template; you'll have to extract the default pdf template first in order to use the CSS styles which are part of the template in order to create a cover page in HTML.

Using a `cover.md` file resolves that requirement because it's processed by the pipeline and the cover HTML file is generated with the CSS styles included from the template.

This is more in-line with how other conceptual documentation files are processed through the pipeline - using a markdown file.

If you want to use HTML files as a type of input, it seems like that HTML file should be part of the template, where other HTML and CSS files exist as part of the pipeline.

superyyrrzz commented:
> I think it should be recommended that cover.md be placed in the pdf folder, like the toc.yml file specific for PDF.

Agree to make this as a convention. You are right that we need not care about site or manifest things here.

> If you want to use HTML files as a type of input, it seems like that HTML file should be part of the template, where other HTML and CSS files exist as part of the pipeline.

👍 I like this idea. The detailed design in DocFX template system could be:

1. user should prepare an input `cover.md` that specifies to use template unique for cover page:
    ```.md
    ---
    documentType: cover
    ---
    # The real cover page title
    ```

2. Add `pdf.default/cover.html.primary.html`, which can contain default styles for cover page.

3. PDF coomand to combine them.

With the default cover page template, user can define cover page easily. If user has many custom style needs as @bitbonk mentinoed, go to custom `cover.html.primary.html`.

icnocop commented:
@superyyrrzz , thank you very much for your detailed review.

I have resolve your review comments in commit https://github.com/dotnet/docfx/pull/4342/commits/34aef22be33c6184a783a8240b40b4db58dc2cc8 as part of this PR.

I also added support for using `cover.html.primary.tmpl` from your suggestion and added it to the default pdf template.

For now, the template simply renders the title in the middle of the page.

However, I didn't understand what you meant by:

> PDF [command] to combine them

Can you clarify that for me please?

Thank you!

superyyrrzz commented:
@icnocop Thank you for the updates! I will review soon.

> PDF [command] to combine them

This just means use the HtmlToPdfConverter to generate PDF, as what your PR already did. Run `docfx.exe` against a config with `build` and `pdf` section is equivalent to run `docfx.exe build` and `docfx.exe pdf`. The latter is what I mean by "pdf command".

superyyrrzz commented on src/Microsoft.DocAsCode.HtmlToPdf/CoverPage/CoverPageProcessor.cs:
```diff
@@ -0,0 +1,85 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
```
Seems no need to add `CoverPageProcessor` and `BuildCoverPage` here. We can leverage existing `ConceptualDocumentProcessor` to process `cover.md` like any other normal Markdown files. The only difference is that `cover.md` user `documentType` to use `cover.html.primary.tmpl` instead of `conceptual.html.primary.html`, which `ConceptualDocumentProcessor` already recognized `documentType` to do this.

superyyrrzz commented on src/Microsoft.DocAsCode.HtmlToPdf/ConvertWrapper.cs:
```diff
@@ -132,6 +132,18 @@ private void ConvertCore(string basePath)
                             htmlModels.Insert(0, new HtmlModel { Title = _pdfOptions.TocTitle, HtmlFilePath = tocPageFilePath });
                         }
+
+                        var coverPage = FindCoverPageInManifest(manifest);
```
Consider the following folder structure, where user want to generate 2 PDFs in one docfx run:
```
pdf
  |- cover.md
  |- toc.yml
pdf1
  |- cover.md
  |- toc.yml
```
Users should expect `pdf/toc.yml` to use `pdf/cover.md`, and `pdf1/toc.yml` to use `pdf1/cover.md`. So I think `FindCoverPageinManifest` should consider `tocFile` page, instead of the first one found globally. It will **only** pick up the `cover.md` alongside, like:
 ```cs
var coverPage = FindSiblingCoverPageInManifest(mainfest, tocFile)
```

superyyrrzz commented on src/docfx.website.themes/pdf.default/cover.html.primary.tmpl:
```diff
@@ -0,0 +1,15 @@
+    margin-left: -200px;
+}
+</style>
+<div class="cover-title">{{{rawTitle}}}</div>
```
We can also add `{{{conceptuall}}}` below. Although it could not be of the ideal format in cover page, users should be happier than finding something missing unexpectedly in output cover.

icnocop commented:
Thanks again for your guidance! 💯 

I've pushed another commit with the changes as you've suggested.

superyyrrzz commented on src/Microsoft.DocAsCode.HtmlToPdf/ConvertWrapper.cs:
```diff
@@ -220,6 +232,15 @@ private IList<ManifestItem> FindTocInManifest(Manifest manifest)
+        {
+            return manifest.Files.SingleOrDefault(f =>
+            {
+                return Path.GetDirectoryName(f.SourceRelativePath).StartsWith(Path.GetDirectoryName(tocFile.SourceRelativePath))
```
Consider 2 folder names like `pdf` and `pdf1`, `StartsWith` is not safe. This line should be more strict like `Path.GetDirectoryName(f.SourceRelativePath) == Path.GetDirectoryName(tocFile.SourceRelativePath)`

superyyrrzz commented on src/Microsoft.DocAsCode.HtmlToPdf/ConvertWrapper.cs:
```diff
@@ -220,6 +232,15 @@ private IList<ManifestItem> FindTocInManifest(Manifest manifest)
+            return manifest.Files.SingleOrDefault(f =>
+            {
+                return Path.GetDirectoryName(f.SourceRelativePath).StartsWith(Path.GetDirectoryName(tocFile.SourceRelativePath))
+                    && (string.Compare("cover.md", Path.GetFileName(f.SourceRelativePath), true) == 0);
```
use `FilePathComparer.OSPlatformSensitiveRelativePathComparer.Equals(..., ...)` This should be case sensitive under Linux.

icnocop replied:
FYI, Windows 10 can be configured for case sensitivity by folder.
https://www.windowscentral.com/how-enable-ntfs-treat-folders-case-sensitive-windows-10

`FilePathComparer.OSPlatformSensitiveRelativePathComparer` will replace back slashes with forward slashes.

superyyrrzz commented on src/Microsoft.DocAsCode.HtmlToPdf/PdfOptions.cs:
```diff
@@ -48,6 +48,11 @@ public class PdfOptions
         /// </summary>
         public OutlineOption OutlineOption { get; set; } = OutlineOption.DefaultOutline;
 
+        /// <summary>
```
After second thoughts, I think the default value of `TocTitle` should be "Table of Contents".

The consideration under my former suggestion is that I do not want users to find there first page changes from "Cover Page" to "Table of Contents" after upgrading DocFX. However, I think this is minor and acceptable, and it can make cover page scenario more smooth. Sorry for this change.

icnocop replied:
It's not a problem. Thanks! :)

superyyrrzz commented on src/Microsoft.DocAsCode.HtmlToPdf/ConvertWrapper.cs:
```diff
@@ -236,8 +236,8 @@ private ManifestItem FindSiblingCoverPageInManifest(Manifest manifest, ManifestI
-                return Path.GetDirectoryName(f.SourceRelativePath).StartsWith(Path.GetDirectoryName(tocFile.SourceRelativePath))
-                    && (string.Compare("cover.md", Path.GetFileName(f.SourceRelativePath), true) == 0);
+                return FilePathComparer.OSPlatformSensitiveRelativePathComparer.Equals(Path.GetDirectoryName(f.SourceRelativePath), Path.GetDirectoryName(tocFile.SourceRelativePath))
+                    && (string.Compare("cover.md", Path.GetFileName(f.SourceRelativePath), ignoreCase: true) == 0);
```
also use `FilePathComparer.OSPlatformSensitiveRelativePathComparer.Equals`

icnocop replied:
What if I use
`"cover.md".Equals(Path.GetFileName(f.SourceRelativePath), StringComparison.OrdinalIgnoreCase)`?

This is similar to how `toc.yml` is compared in other places:
`"toc.yml".Equals(Path.GetFileName(x.File), StringComparison.OrdinalIgnoreCase)`

superyyrrzz replied:
One corner case is under Linux, the folder may exist both `cover.md` and `COVER.md`. Finding multiple matches will fail `SingleOrDefault`. This is a real corner case, so the current implementation does make sense. Never mind 😄

icnocop replied:
Oh, you're right! I didn't think of that. 😄

superyyrrzz commented:
@icnocop Thank you for contribution for this amazing feature!

icnocop commented:
You're welcome. Thank you @superyyrrzz for the quick feedback! 🎉
